chains/substrate/listener.go

```diff
@@ -107,32 +107,57 @@ func (l *listener) pollBlocks() error {
 				return nil
 			}
 
+			// Get finalized block hash
+			finalizedHash, err := l.conn.api.RPC.Chain.GetFinalizedHead()
+			if err != nil {
+				l.log.Error("Failed to fetch finalized hash", "err", err)
+				retry--
+				time.Sleep(BlockRetryInterval)
+				continue
+			}
+
+			// Get finalized block header
+			finalizedHeader, err := l.conn.api.RPC.Chain.GetHeader(finalizedHash)
+			if err != nil {
+				l.log.Error("Failed to fetch finalized header", "err", err)
+				retry--
+				time.Sleep(BlockRetryInterval)
+				continue
+			}
+
+			// Sleep if the block we want comes after the most recently finalized block
+			if currentBlock > uint64(finalizedHeader.Number) {
```

Comment:
Should we add another condition to check if the `currentBlock` comes after the most recently finalized block by more than our block retry limit?

Reply:
This case doesn't decrement the retry count. We always retry if the block is not ready, I think progress is somewhat guaranteed.

Reply:
kk that's fair, even if we're facing adversarial network conditions eventually another block will be finalized.